vermouth/processors/do_mapping.py

```diff
@@ -245,54 +263,88 @@ def do_mapping(molecule, mappings, to_ff, attribute_keep=()):
-    for match, name, mapping in sorted(all_matches, key=lambda x: min(x[0].keys())):
-        blocks_per_atom.update(match.keys())
+    for mol_to_block, blocks_to, references in sorted(all_matches, key=lambda x: min(x[0].keys())):
         if graph_out.nrexcl is None:
-            graph_out.nrexcl = mapping.block_to.nrexcl
+            graph_out.nrexcl = blocks_to.nrexcl
         try:
             # merge_molecule will return a dict mapping the node keys of the
             # added block to the ones in graph_out
-            block_to_out = graph_out.merge_molecule(mapping.block_to)
+            # FIXME: Issue #154 lives here.
+            block_to_out = graph_out.merge_molecule(blocks_to)
         except ValueError:
             # This probably means the nrexcl of the block is different from the
             # others. This means the user messed up their data. Or there are
             # different forcefields in the same forcefield folder...
-            LOGGER.exception('Residue {} is not compatible with the others',
-                             name, type='inconsistent-data')
+            LOGGER.exception('Residue(s) {} is not compatible with the others',
+                             set(nx.get_node_attributes(blocks_to, 'resname').values()),
+                             type='inconsistent-data')
             raise
-        block_to_mol = {v: k for k, v in match.items()}
-        for to_idx, from_idxs in mapping.mapping.items():
-            # Some bookkeeping with indices.
-            out_idx = block_to_out[to_idx]
-            mol_idxs = [block_to_mol[from_idx] for from_idx in from_idxs]
-            for mol_idx in mol_idxs:
-                mol_to_out[mol_idx].append(out_idx)
-
-            # Keep track of what bead comes from where
-            subgraph = molecule.subgraph(mol_idxs)
-            graph_out.nodes[out_idx]['graph'] = subgraph
-            weights = {block_to_mol[from_idx]: mapping.weights[to_idx][from_idx]
-                       for from_idx in from_idxs}
-            graph_out.nodes[out_idx]['mapping_weights'] = weights
+        # overlapping_mappings does not have to be a dict, since the values in
+        # block_to_out are guaranteed to be unique in graph_out. So we can look
+        # them up in mol_to_out
+        overlapping_mappings.update(set(mol_to_block.keys()) & set(mol_to_out.keys()))
+        for mol_idx in mol_to_block:
+            for block_idx, weight in mol_to_block[mol_idx].items():
+                out_idx = block_to_out[block_idx]
+                mol_to_out[mol_idx][out_idx] = weight
+                out_to_mol[out_idx][mol_idx] = weight
+
+        mapped_block_idxs = {idx for mol_idx in mol_to_block for idx in mol_to_block[mol_idx]}
+        for spawned in set(blocks_to.nodes) - mapped_block_idxs:
+            # These nodes come from "nowhere", so, let's pretend they come from
+            # all nodes in the block. This helps with setting attributes such
+            # as 'chain'
```

Comment:
How does this impact connectivity?

Reply:
nodes in none_to_one_mappings get explicitly removed from making edges to neighbouring residues automatically, since that would indeed result in problems.